File: treedex/tree_builder.py
def toc_to_sections(toc: list[dict]) -> list[dict]:
    """Convert ToC entries ``{level, title, physical_index}`` into flat
    sections with hierarchical ``structure`` numbering (``"1"``, ``"1.2"``,
    ``"1.2.3"``).
    """
    counters: dict[int, int] = {}
    sections: list[dict] = []

    for entry in toc:
        level = entry["level"]

        # Reset deeper counters
        for k in list(counters):
            if k > level:
                del counters[k]

        counters[level] = counters.get(level, 0) + 1

        parts = [str(counters.get(l, 1)) for l in range(1, level + 1)]

        sections.append({
            "structure": ".".join(parts),
            "title": entry["title"],
            "physical_index": entry["physical_index"],
        })

    return sections


def repair_orphans(flat_list: list[dict]) -> list[dict]:
    """Repair orphaned sections by inserting synthetic parent nodes.

    If ``"2.3.1"`` exists but ``"2.3"`` doesn't, a synthetic ``"2.3"`` node
    is inserted so that ``list_to_tree`` can build the correct hierarchy.
    """
    known = {item["structure"] for item in flat_list}
    inserts: list[dict] = []

    for item in flat_list:
        parts = item["structure"].split(".")
        for depth in range(1, len(parts)):
            ancestor = ".".join(parts[:depth])
            if ancestor not in known:
                inserts.append({
                    "structure": ancestor,
                    "title": f"Section {ancestor}",
                    "physical_index": item["physical_index"],
                })
                known.add(ancestor)

    if inserts:
        combined = flat_list + inserts
        combined.sort(key=lambda x: [int(p) for p in x["structure"].split(".")])
        return combined

    return flat_list


def list_to_tree(flat_list: list[dict]) -> list[dict]:
    """Convert a flat list with `structure` fields into a hierarchical tree.

    Each item must have a `structure` field like "1", "1.1", "1.2.3".
    Parent of "1.2.3" is "1.2", parent of "1.2" is "1", "1" is root.
    Output nodes get a `nodes: []` field for children.
    """
    nodes_by_structure = {}
    roots = []

    for item in flat_list:
        node = {**item, "nodes": []}
        structure = node["structure"]
        nodes_by_structure[structure] = node

        parts = structure.rsplit(".", 1)
        if len(parts) == 1:
            roots.append(node)
        else:
            parent_structure = parts[0]
            parent = nodes_by_structure.get(parent_structure)
            if parent is not None:
                parent["nodes"].append(node)
            else:
                roots.append(node)

    return roots


def _assign_ranges(nodes: list[dict], boundary_end: int):
    """Recursively assign start_index and end_index to nodes."""
    for i, node in enumerate(nodes):
        node["start_index"] = node.get("physical_index", 0)

        if i + 1 < len(nodes):
            node["end_index"] = nodes[i + 1].get("physical_index", 0) - 1
        else:
            node["end_index"] = boundary_end

        # Clamp: end must be >= start (sections on the same page)
        if node["end_index"] < node["start_index"]:
            node["end_index"] = node["start_index"]

        if node.get("nodes"):
            _assign_ranges(node["nodes"], node["end_index"])


def assign_page_ranges(tree: list[dict], total_pages: int) -> list[dict]:
    """Set start_index and end_index on each node.

    - start_index = node's physical_index
    - end_index = next sibling's physical_index - 1, or parent's end,
      or total_pages - 1 for the last root node
    """
    _assign_ranges(tree, total_pages - 1)
    return tree


def assign_node_ids(tree: list[dict]) -> list[dict]:
    """DFS traversal, assigns sequential IDs: '0001', '0002', etc."""
    counter = [0]

    def _walk(nodes):
        for node in nodes:
            counter[0] += 1
            node["node_id"] = f"{counter[0]:04d}"
            _walk(node.get("nodes", []))

    _walk(tree)
    return tree


def find_large_nodes(
    tree: list[dict],
    max_pages: int = 10,
    max_tokens: int = 20000,
    pages: list[dict] | None = None,
) -> list[dict]:
    """Return nodes that exceed page or token thresholds."""
    large = []

    def _walk(nodes):
        for node in nodes:
            start = node.get("start_index", 0)
            end = node.get("end_index", 0)
            page_count = end - start + 1

            is_large = page_count > max_pages

            if not is_large and pages is not None:
                token_sum = sum(
                    p["token_count"]
                    for p in pages
                    if start <= p["page_num"] <= end
                )
                is_large = token_sum > max_tokens

            if is_large:
                large.append(node)

            _walk(node.get("nodes", []))

    _walk(tree)
    return large


def embed_text_in_tree(tree: list[dict], pages: list[dict]) -> list[dict]:
    """Add `text` field to each node by concatenating page text for its range."""

    def _walk(nodes):
        for node in nodes:
            start = node.get("start_index", 0)
            end = node.get("end_index", 0)
            node["text"] = "\n".join(
                p["text"] for p in pages if start <= p["page_num"] <= end
            )
            _walk(node.get("nodes", []))

    _walk(tree)
    return tree

File: src/tree-builder.ts
/** Tree construction utilities. */

import type { TreeNode, Page } from "./types.js";

export interface TocEntry {
  level: number;
  title: string;
  physical_index: number;
}

/**
 * Convert ToC entries `{level, title, physical_index}` into flat sections
 * with hierarchical `structure` numbering (`"1"`, `"1.2"`, `"1.2.3"`).
 */
export function tocToSections(
  toc: TocEntry[],
): Array<{ structure: string; title: string; physical_index: number }> {
  const counters: Record<number, number> = {};
  const sections: Array<{
    structure: string;
    title: string;
    physical_index: number;
  }> = [];

  for (const entry of toc) {
    const level = entry.level;

    // Reset deeper counters
    for (const k of Object.keys(counters)) {
      if (Number(k) > level) {
        delete counters[Number(k)];
      }
    }

    counters[level] = (counters[level] ?? 0) + 1;

    const parts: string[] = [];
    for (let l = 1; l <= level; l++) {
      parts.push(String(counters[l] ?? 1));
    }

    sections.push({
      structure: parts.join("."),
      title: entry.title,
      physical_index: entry.physical_index,
    });
  }

  return sections;
}

/**
 * Repair orphaned sections by inserting synthetic parent nodes.
 *
 * If `"2.3.1"` exists but `"2.3"` doesn't, a synthetic `"2.3"` node is
 * inserted so that `listToTree` can build the correct hierarchy.
 */
export function repairOrphans(
  flatList: Array<{
    structure: string;
    title: string;
    physical_index: number;
    [key: string]: unknown;
  }>,
): Array<{
  structure: string;
  title: string;
  physical_index: number;
  [key: string]: unknown;
}> {
  const known = new Set(flatList.map((item) => item.structure));
  const inserts: Array<{
    structure: string;
    title: string;
    physical_index: number;
  }> = [];

  for (const item of flatList) {
    const parts = item.structure.split(".");
    for (let depth = 1; depth < parts.length; depth++) {
      const ancestor = parts.slice(0, depth).join(".");
      if (!known.has(ancestor)) {
        inserts.push({
          structure: ancestor,
          title: `Section ${ancestor}`,
          physical_index: item.physical_index,
        });
        known.add(ancestor);
      }
    }
  }

  if (inserts.length > 0) {
    const combined = [...flatList, ...inserts];
    combined.sort((a, b) => {
      const ap = a.structure.split(".").map(Number);
      const bp = b.structure.split(".").map(Number);
      for (let i = 0; i < Math.max(ap.length, bp.length); i++) {
        const av = ap[i] ?? 0;
        const bv = bp[i] ?? 0;
        if (av !== bv) return av - bv;
      }
      return 0;
    });
    return combined;
  }

  return flatList;
}

/**
 * Convert a flat list with `structure` fields into a hierarchical tree.
 *
 * Each item must have a `structure` field like "1", "1.1", "1.2.3".
 * Parent of "1.2.3" is "1.2", parent of "1.2" is "1", "1" is root.
 * Output nodes get a `nodes: []` field for children.
 */
export function listToTree(
  flatList: Array<{
    structure: string;
    title: string;
    physical_index: number;
    [key: string]: unknown;
  }>,
): TreeNode[] {
  const nodesByStructure: Record<string, TreeNode> = {};
  const roots: TreeNode[] = [];

  for (const item of flatList) {
    const node: TreeNode = { ...item, nodes: [] };
    const structure = node.structure;
    nodesByStructure[structure] = node;

    const parts = structure.split(".");
    if (parts.length === 1) {
      roots.push(node);
    } else {
      const parentStructure = parts.slice(0, -1).join(".");
      const parent = nodesByStructure[parentStructure];
      if (parent !== undefined) {
        parent.nodes.push(node);
      } else {
        roots.push(node);
      }
    }
  }

  return roots;
}

function assignRanges(nodes: TreeNode[], boundaryEnd: number): void {
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    node.start_index = node.physical_index ?? 0;

    if (i + 1 < nodes.length) {
      node.end_index = (nodes[i + 1].physical_index ?? 0) - 1;
    } else {
      node.end_index = boundaryEnd;
    }

    // Clamp: end must be >= start (sections on the same page)
    if (node.end_index < node.start_index) {
      node.end_index = node.start_index;
    }

    if (node.nodes.length > 0) {
      assignRanges(node.nodes, node.end_index);
    }
  }
}

/**
 * Set start_index and end_index on each node.
 *
 * - start_index = node's physical_index
 * - end_index = next sibling's physical_index - 1, or parent's end,
 *   or total_pages - 1 for the last root node
 */
export function assignPageRanges(
  tree: TreeNode[],
  totalPages: number,
): TreeNode[] {
  assignRanges(tree, totalPages - 1);
  return tree;
}

/** DFS traversal, assigns sequential IDs: '0001', '0002', etc. */
export function assignNodeIds(tree: TreeNode[]): TreeNode[] {
  let counter = 0;

  function walk(nodes: TreeNode[]): void {
    for (const node of nodes) {
      counter++;
      node.node_id = String(counter).padStart(4, "0");
      walk(node.nodes);
    }
  }

  walk(tree);
  return tree;
}

/** Return nodes that exceed page or token thresholds. */
export function findLargeNodes(
  tree: TreeNode[],
  options: {
    maxPages?: number;
    maxTokens?: number;
    pages?: Page[] | null;
  } = {},
): TreeNode[] {
  const { maxPages = 10, maxTokens = 20000, pages = null } = options;
  const large: TreeNode[] = [];

  function walk(nodes: TreeNode[]): void {
    for (const node of nodes) {
      const start = node.start_index ?? 0;
      const end = node.end_index ?? 0;
      const pageCount = end - start + 1;

      let isLarge = pageCount > maxPages;

      if (!isLarge && pages !== null) {
        const tokenSum = pages
          .filter((p) => p.page_num >= start && p.page_num <= end)
          .reduce((sum, p) => sum + p.token_count, 0);
        isLarge = tokenSum > maxTokens;
      }

      if (isLarge) {
        large.push(node);
      }

      walk(node.nodes);
    }
  }

  walk(tree);
  return large;
}

/** Add `text` field to each node by concatenating page text for its range. */
export function embedTextInTree(
  tree: TreeNode[],
  pages: Page[],
): TreeNode[] {
  function walk(nodes: TreeNode[]): void {
    for (const node of nodes) {
      const start = node.start_index ?? 0;
      const end = node.end_index ?? 0;
      node.text = pages
        .filter((p) => p.page_num >= start && p.page_num <= end)
        .map((p) => p.text)
        .join("\n");
      walk(node.nodes);
    }
  }

  walk(tree);
  return tree;
}
